### PR TITLE
External transports

### PR DESCRIPTION
**Pull request by pablogs9**
This PR adds external transports to Micro-XRCE-DDS-Client build

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -66,6 +55,21 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
...
+set(BIN_INSTALL_DIR     ${CMAKE_INSTALL_BINDIR}     CACHE PATH "Installation directory for binaries.")
+set(INCLUDE_INSTALL_DIR ${CMAKE_INSTALL_INCLUDEDIR} CACHE PATH "Installation directory for C headers.")
+set(LIB_INSTALL_DIR     ${CMAKE_INSTALL_LIBDIR}     CACHE PATH "Installation directory for libraries.")
+set(DATA_INSTALL_DIR    ${CMAKE_INSTALL_DATADIR}    CACHE PATH "Installation directory for data.")
+if(WIN32)
+    set(LICENSE_INSTALL_DIR . CACHE PATH "Installation directory for licenses.")
+else()
+    set(LICENSE_INSTALL_DIR ${DATA_INSTALL_DIR}/${PROJECT_NAME} CACHE PATH "Installation directory for licenses.")
+endif()
+
```
Comment (julianbermudez) [inline CMakeLists.txt]: It seems that you were working in develop branch before the master merge. Please try to maintain the develop branch changes.

File: CMakeLists.txt [hunk 2]
```diff
@@ -102,6 +106,64 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
...
+                src/c/profile/transport/${TT}/${TT}_transport_linux.c
+                )
+        elseif(PLATFORM_EXTERNAL)
+            set(${TRANS_TYPE}_SRCS
+                src/c/profile/transport/${TT}/${TT}_transport.c
+                ${PLATFORM_EXTERNAL_TRANSPORT}
+            )
+        endif()
+    endif()
+endforeach()
```
Comment (julianbermudez) [inline CMakeLists.txt]: This approach changed, now we add the sources when we set the `SRCS`variable. Take a look to line 199.

File: CMakeLists.txt [hunk 3]
```diff
@@ -157,8 +219,6 @@ set_common_compile_options(${PROJECT_NAME})
 set_target_properties(${PROJECT_NAME} PROPERTIES
     VERSION
         ${PROJECT_VERSION}
-    SOVERSION
-        ${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}
```
Comment (julianbermudez) [inline CMakeLists.txt]: Please don't remove it, the soversion is important.

File: CMakeLists.txt [hunk 4]
```diff
@@ -206,6 +266,13 @@ target_include_directories(${PROJECT_NAME}
         $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src/c>
     )
 
+if(PLATFORM_EXTERNAL)    
+    target_include_directories(${PROJECT_NAME} PUBLIC
+        ${PLATFORM_EXTERNAL_TRANSPORT_HEADER_DIRECTORY}
+        $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src/c/profile/transport/serial>
```
Comment (julianbermudez) [inline CMakeLists.txt]: Why do you add an src file as an include directory?

File: CMakeLists.txt [hunk 5]
```diff
@@ -235,8 +302,6 @@ endif()
 if(UCLIENT_BUILD_EXAMPLES)
     add_subdirectory(examples/PublishHelloWorld)
     add_subdirectory(examples/SubscribeHelloWorld)
-    add_subdirectory(examples/PublishHelloWorldP2P)
-    add_subdirectory(examples/SubscribeHelloWorldP2P)
```
Comment (julianbermudez) [inline CMakeLists.txt]: Please don't remove P2P examples.

File: include/uxr/client/config.h.in [hunk 6]
```diff
@@ -29,6 +29,9 @@
 #cmakedefine PLATFORM_NAME_LINUX
 #cmakedefine PLATFORM_NAME_WINDOWS
 #cmakedefine PLATFORM_NAME_NUTTX
+#cmakedefine PLATFORM_EXTERNAL
+
+#cmakedefine PLATFORM_EXTERNAL_TRANSPORT_HEADER_FILENAME <uxr/client/profile/transport/serial/@PLATFORM_EXTERNAL_TRANSPORT_HEADER_FILENAME@>
```
Comment (julianbermudez) [inline include/uxr/client/config.h.in]: Maybe the `PLATFORM_` prefix is not necessary, what do you think?

File: include/uxr/client/transport.h [hunk 7]
```diff
@@ -46,6 +46,8 @@
 #elif defined(PLATFORM_NAME_WINDOWS)
 #elif defined(PLATFORM_NAME_NUTTX)
 #include <uxr/client/profile/transport/serial/serial_transport_linux.h>
+#elif defined(PLATFORM_EXTERNAL)
+#include PLATFORM_EXTERNAL_TRANSPORT_HEADER_FILENAME
```
Comment (julianbermudez) [inline include/uxr/client/transport.h]: Same comment as before.

File: CMakeLists.txt [hunk 8]
```diff
@@ -102,6 +106,64 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
...
         set(PLATFORM_NAME_NUTTX ON)
+        set(TRANSPORT_TYPES "UDP" "TCP" "SERIAL")
+    endif()
+endif()
+
+# Check external transport.
+SET(PLATFORM_EXTERNAL_TRANSPORT_HEADER "" CACHE PATH "External transports header file")
+SET(PLATFORM_EXTERNAL_TRANSPORT "" CACHE PATH "External transports code file")
+
+if( PLATFORM_EXTERNAL_TRANSPORT)
```
Comment (julianbermudez) [inline CMakeLists.txt]: Please remove space.

File: CMakeLists.txt [hunk 9]
```diff
@@ -102,6 +106,64 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
...
+    set(PLATFORM_EXTERNAL ON)
+
+    set(TRANSPORT_TYPES "SERIAL")
```
Comment (julianbermudez) [inline CMakeLists.txt]: Could we prepare it for `UDP` and `TCP` transports?

File: CMakeLists.txt [hunk 10]
```diff
@@ -102,6 +106,64 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
...
+
+    get_filename_component(PLATFORM_EXTERNAL_TRANSPORT_HEADER_FILENAME ${PLATFORM_EXTERNAL_TRANSPORT_HEADER} NAME)
+    get_filename_component(PLATFORM_EXTERNAL_TRANSPORT_FILENAME ${PLATFORM_EXTERNAL_TRANSPORT} NAME)
+    get_filename_component(PLATFORM_EXTERNAL_TRANSPORT_HEADER_DIRECTORY ${PLATFORM_EXTERNAL_TRANSPORT_HEADER} DIRECTORY)
```
Comment (julianbermudez) [inline CMakeLists.txt]: Maybe we can remove the `PLATFORM_` prefix. What do you think?
On the other hand, now we use lowercase name with dash prefix for internal variables.

File: CMakeLists.txt (hunk 1 above)
Comment (pablogs9) [reply CMakeLists.txt]: Ok, just a rebasing mistake.

File: CMakeLists.txt (hunk 2 above)
Comment (pablogs9) [reply CMakeLists.txt]: New approach used in external transports.

File: CMakeLists.txt (hunk 3 above)
Comment (pablogs9) [reply CMakeLists.txt]: Ok, just a rebasing mistake.

File: CMakeLists.txt (hunk 4 above)
Comment (pablogs9) [reply CMakeLists.txt]: I need to have access to serial_transport_internal.h in order to provide an external transport. This file is in folder /src/c/profile/transport/serial.

File: CMakeLists.txt (hunk 5 above)
Comment (pablogs9) [reply CMakeLists.txt]: Sorry, just a rebasing mistake.

File: include/uxr/client/config.h.in (hunk 6 above)
Comment (pablogs9) [reply include/uxr/client/config.h.in]: `PLATFORM_` prefix removed.

File: include/uxr/client/transport.h (hunk 7 above)
Comment (pablogs9) [reply include/uxr/client/transport.h]: `PLATFORM_` prefix removed.

File: CMakeLists.txt (hunk 8 above)
Comment (pablogs9) [reply CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 10 above)
Comment (pablogs9) [reply CMakeLists.txt]: `PLATFORM_` prefix removed.
Where can I find a lowercase name example along this file?

File: CMakeLists.txt (hunk 9 above)
Comment (pablogs9) [reply CMakeLists.txt]: Maybe with a `-DEXTERNAL_TYPE` flag?

File: CMakeLists.txt [hunk 11]
```diff
@@ -105,6 +105,17 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Generic")
     endif()
 endif()
 
+# Check external transport.
+SET(EXTERNAL_TRANSPORT_HEADER ""   CACHE PATH "External transports header file")
+SET(EXTERNAL_TRANSPORT ""          CACHE PATH "External transports code file")
```
Comment (julianbermudez) [inline CMakeLists.txt]: Maybe you can use `EXTERNAL_TRANSPORT_SRC` or `EXTERNAL_TRANSPORT_SOURCE` instead.

Comment (julianbermudez) [inline CMakeLists.txt]: Please use lowercase for CMake functions.

File: CMakeLists.txt [hunk 12]
```diff
@@ -133,18 +144,19 @@ set(SRCS
     $<$<AND:$<BOOL:${PROFILE_DISCOVERY}>,$<PLATFORM_ID:Linux>>:src/c/profile/discovery/transport/udp_transport_datagram_linux.c>
     $<$<AND:$<BOOL:${PROFILE_DISCOVERY}>,$<PLATFORM_ID:Windows>>:src/c/profile/discovery/transport/udp_transport_datagram_windows.c>
     $<$<AND:$<BOOL:${PROFILE_DISCOVERY}>,$<PLATFORM_ID:Generic>,$<STREQUAL:${PLATFORM_NAME},nuttx>>:src/c/profile/discovery/transport/udp_transport_datagram_linux.c>
-    $<$<BOOL:${PROFILE_UDP_TRANSPORT}>:src/c/profile/transport/ip/udp/udp_transport.c>
+    $<$<BOOL:${PROFILE_UDP_TRANSPORT}>:src/c/profile/transport/ip/udp/udp_transport.c>
```
Comment (julianbermudez) [inline CMakeLists.txt]: Please remove spaces at the end of the line.

File: CMakeLists.txt [hunk 13]
```diff
@@ -133,18 +144,19 @@ set(SRCS
...
     $<$<AND:$<BOOL:${PROFILE_UDP_TRANSPORT}>,$<PLATFORM_ID:Generic>,$<STREQUAL:${PLATFORM_NAME},nuttx>>:src/c/profile/transport/ip/udp/udp_transport_linux.c>
     $<$<BOOL:${PROFILE_TCP_TRANSPORT}>:src/c/profile/transport/ip/tcp/tcp_transport.c>
     $<$<AND:$<BOOL:${PROFILE_TCP_TRANSPORT}>,$<PLATFORM_ID:Linux>>:src/c/profile/transport/ip/tcp/tcp_transport_linux.c>
     $<$<AND:$<BOOL:${PROFILE_TCP_TRANSPORT}>,$<PLATFORM_ID:Windows>>:src/c/profile/transport/ip/tcp/tcp_transport_windows.c>
     $<$<AND:$<BOOL:${PROFILE_TCP_TRANSPORT}>,$<PLATFORM_ID:Generic>,$<STREQUAL:${PLATFORM_NAME},nuttx>>:src/c/profile/transport/ip/tcp/tcp_transport_linux.c>
-    $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<PLATFORM_ID:Linux>>:src/c/profile/transport/serial/serial_transport.c>
-    $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<PLATFORM_ID:Linux>>:src/c/profile/transport/serial/serial_protocol.c>
+    $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<OR:$<PLATFORM_ID:Linux>,$<PLATFORM_ID:Generic>>>:src/c/profile/transport/serial/serial_transport.c>
+    $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<OR:$<PLATFORM_ID:Linux>,$<PLATFORM_ID:Generic>>>:src/c/profile/transport/serial/serial_protocol.c>
```
Comment (julianbermudez) [inline CMakeLists.txt]: Thank you so much for clearing my trash ;-)

File: CMakeLists.txt [hunk 14]
```diff
@@ -105,6 +105,17 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Generic")
     endif()
 endif()
 
+# Check external transport.
+SET(EXTERNAL_TRANSPORT_HEADER ""   CACHE PATH "External transports header file")
+SET(EXTERNAL_TRANSPORT ""          CACHE PATH "External transports code file")
+
+if(EXTERNAL_TRANSPORT)
```
Comment (julianbermudez) [inline CMakeLists.txt]: What happen if some give you only one, `EXTERNAL_TRANSPORT` or `EXTERNAL_TRANSPORT_HEADER`?

File: CMakeLists.txt [hunk 15]
```diff
@@ -105,6 +105,17 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Generic")
...
+    set(EXTERNAL ON)
```
Comment (julianbermudez) [inline CMakeLists.txt]: Please use lowercase for internal variables.

File: CMakeLists.txt [hunk 16]
```diff
@@ -291,6 +305,16 @@ install(
...
     )
 
+# Install external transport header
+if(EXTERNAL)
+    install(
+        FILES
+            ${EXTERNAL_TRANSPORT_HEADER}
+        DESTINATION
+            ${INCLUDE_INSTALL_DIR}/uxr/client/profile/transport/serial
+    )
```
Comment (julianbermudez) [inline CMakeLists.txt]: Please, insert 4 space.

File: CMakeLists.txt [hunk 17]
```diff
@@ -202,6 +214,8 @@ target_include_directories(${PROJECT_NAME}
         $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
         $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/include>
         $<INSTALL_INTERFACE:include>
+        $<$<BOOL:${EXTERNAL}>:${EXTERNAL_TRANSPORT_HEADER_DIRECTORY}>
+        $<$<BOOL:${EXTERNAL}>:${PROJECT_SOURCE_DIR}/src/c/profile/transport/serial>
```
Comment (julianbermudez) [inline CMakeLists.txt]: Why don't you use neither `BUILD_INTERFACE` nor `INSTALL_INTERFACE`?

File: CMakeLists.txt (hunk 11 above)
Comment (pablogs9) [reply CMakeLists.txt]: Done

Comment (pablogs9) [reply CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 12 above)
Comment (pablogs9) [reply CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 14 above)
Comment (pablogs9) [reply CMakeLists.txt]: Now I check both at line 112

File: CMakeLists.txt (hunk 15 above)
Comment (pablogs9) [reply CMakeLists.txt]: `EXTERNAL` moved to `external`

File: CMakeLists.txt (hunk 16 above)
Comment (pablogs9) [reply CMakeLists.txt]: Where?

File: CMakeLists.txt (hunk 17 above)
Comment (pablogs9) [reply CMakeLists.txt]: Done with `BUILD_INTERFACE`

File: CMakeLists.txt [hunk 18]
```diff
@@ -291,6 +305,16 @@ install(
...
     )
 
+# Install external transport header
+if(external)
+    install(
+        FILES
+            ${EXTERNAL_TRANSPORT_HEADER}
+        DESTINATION
+            ${INCLUDE_INSTALL_DIR}/uxr/client/profile/transport/serial
+    )
```
Comment (julianbermudez) [inline CMakeLists.txt]:
```suggestion
        )
```

File: CMakeLists.txt [hunk 19]
```diff
@@ -105,6 +105,17 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Generic")
     endif()
 endif()
 
+# Check external transport.
+set(EXTERNAL_TRANSPORT_HEADER ""    CACHE PATH "External transports header file")
+set(EXTERNAL_TRANSPORT_SRC ""       CACHE PATH "External transports code file")
+
+if(EXTERNAL_TRANSPORT_SRC AND EXTERNAL_TRANSPORT_HEADER)
+    set(external ON)
```
Comment (julianbermudez) [inline CMakeLists.txt]:
```suggestion
    set(_external_transport ON)
```

File: CMakeLists.txt [hunk 20]
```diff
@@ -141,10 +152,11 @@ set(SRCS
...
     $<$<AND:$<BOOL:${PROFILE_TCP_TRANSPORT}>,$<PLATFORM_ID:Generic>,$<STREQUAL:${PLATFORM_NAME},nuttx>>:src/c/profile/transport/ip/tcp/tcp_transport_linux.c>
-    $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<PLATFORM_ID:Linux>>:src/c/profile/transport/serial/serial_transport.c>
-    $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<PLATFORM_ID:Linux>>:src/c/profile/transport/serial/serial_protocol.c>
+    $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<OR:$<PLATFORM_ID:Linux>,$<PLATFORM_ID:Generic>>>:src/c/profile/transport/serial/serial_transport.c>
+    $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<OR:$<PLATFORM_ID:Linux>,$<PLATFORM_ID:Generic>>>:src/c/profile/transport/serial/serial_protocol.c>
     $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<PLATFORM_ID:Linux>>:src/c/profile/transport/serial/serial_transport_linux.c>
     $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<PLATFORM_ID:Generic>,$<STREQUAL:${PLATFORM_NAME},nuttx>>:src/c/profile/transport/serial/serial_transport_linux.c>
+    $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<PLATFORM_ID:Generic>,$<BOOL:${external}>>:${EXTERNAL_TRANSPORT_SRC}>
```
Comment (julianbermudez) [inline CMakeLists.txt]:
```suggestion
    $<$<AND:$<BOOL:${PROFILE_SERIAL_TRANSPORT}>,$<PLATFORM_ID:Generic>,$<BOOL:${_external_transport}>>:${EXTERNAL_TRANSPORT_SRC}>
```

File: CMakeLists.txt [hunk 21]
```diff
@@ -202,6 +214,8 @@ target_include_directories(${PROJECT_NAME}
         $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
         $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/include>
         $<INSTALL_INTERFACE:include>
+        $<$<AND:$<BOOL:${BUILD_INTERFACE}>,$<BOOL:${external}>>:${EXTERNAL_TRANSPORT_HEADER_DIRECTORY}>
```
Comment (julianbermudez) [inline CMakeLists.txt]:
```suggestion
        $<$<AND:$<BOOL:${BUILD_INTERFACE}>,$<BOOL:${_external_transport}>>:${EXTERNAL_TRANSPORT_HEADER_DIRECTORY}>
```

File: CMakeLists.txt [hunk 22]
```diff
@@ -202,6 +214,8 @@ target_include_directories(${PROJECT_NAME}
...
+        $<$<AND:$<BOOL:${BUILD_INTERFACE}>,$<BOOL:${external}>>:${PROJECT_SOURCE_DIR}/src/c/profile/transport/serial>
```
Comment (julianbermudez) [inline CMakeLists.txt]:
```suggestion
        $<$<AND:$<BOOL:${BUILD_INTERFACE}>,$<BOOL:${_external_transport}>>:${PROJECT_SOURCE_DIR}/src/c/profile/transport/serial>
```

File: CMakeLists.txt [hunk 23]
```diff
@@ -291,6 +305,16 @@ install(
         PATTERN "*.h"
     )
 
+# Install external transport header
+if(external)
```
Comment (julianbermudez) [inline CMakeLists.txt]:
```suggestion
if(_external_transport)
```

File: CMakeLists.txt (hunk 22 above)
Comment (julianbermudez) [reply CMakeLists.txt]: In case you need include a header located in _src_ folder, we should move it to _include_ directory.

**Pull request merged**